### ISSUE TITLE
Unsafe conversion from double to long results in `:type-mismatch` error.

### ISSUE THREAD

**Issue by genmeblog** [clj-kondo/clj-kondo#2824]
- [x] I have read the [Clojure etiquette](https://clojure.org/community/etiquette) and will respect it when communicating on this platform.

**version**

`v2026.04.15`

**macro usage**

no

**platform**

WSL2 Ubuntu

**editor**

Emacs/Flycheck

**problem**

When function accepts primitive long, coersion from double yields a type mismatch error.  It's of course unsafe promotion but it's valid in certain cases.

**repro**

```clojure
(defn my-log-wrapper
  ^double [^long v] (Math/log v))

(my-log-wrapper 2.1)
```

Results in:
`Expected: long, received: double`

**expected behavior**

I think it shouldn't be considered as an error - just a warning. Maybe such primitive, unsafe conversion could be validated by a separate linter. Or maybe `:type-mismatch` can get an option to exclude certain type pairs?

**Comment by borkdude** [comment]
If passing a double where a long is expected is ok, we can fix this in the type system itself. Not sure if this a good idea in general?

**Comment by genmeblog** [comment]
True, in general such implicit conversion can be error prone. But this case deserves a warning not an error. I have some combination of inlined function, java interop and primitive hints which fools clj-kondo and I get false positives.

**Comment by borkdude** [comment]
> It's of course unsafe promotion but it's valid in certain cases.

clj-kondo doesn't know about "certain cases" unless you can specify which certain cases very clearly. Not sure what action I should take right now unless you can explain it to me like I'm five years old perhaps.

Overly specific false positives can be gotten rid of with `#_:clj-kondo/ignore` before the warning.
